### PR TITLE
Enable partial stacking at Batch level

### LINKED ISSUE #90: Batch to Numpy compatibility
The current implementation of `__len__` and `__getitem__` is kind of weird because it is infinitely recursive: one can do `Batch()[0]` and it does not throw an exception but rather return the same object `Batch()`. This is an issue because it breaks the compatibility with standard numpy methods such as `np.mean`, `np.max` ... which should work otherwise. It would be easy to change this behavior, but many policy algorithm relies on it.  Calling such a numpy method simply hangs infinitely, for example using `np.asarray(Batch())`, and so does `[o for o in Batch(a=[1.0])]` which is even worst.

I opened a PR implementing the desired behavior, but I need help to update accordingly the policy algorithms.

______

Here is what you get using the proposed implementation.

Input:
```
np.mean(Batch(a=np.array([[1.0, 2.0], [3.0, 4.0]]), b=np.array([5.0, 6.0])))
```
Output:
```
Batch(
    a: array([2., 3.]),
    b: 5.5,
)
```

### LINKED ISSUE #96: Batch & Buffer profiling
Related to #95

### PR DESCRIPTION
**Pull request by duburcqa**
Allow stacking of inconsistent batches:
```
In [1]: from tianshou.data import Batch, to_numpy ; import numpy as np ; import torch                                                                                                                       
In [2]: Batch(np.array([{'is_success': True, 'reward': {'done': 0.0001, 'direction': -0.11480981}}, 
   ...:                 {'is_success': False, 'reward': {'done': 0.0001}}], dtype=object))                                                                                                                  
Out[2]: 
Batch(
    is_success: array([ True, False]),
    reward: Batch(
                done: array([0.0001, 0.0001]),
                direction: array([-0.11480981,  0.        ]),
            ),
)
```

Before that, the result was unexpected (no failure and different result depending of Batch order).

As you may notice, I also replaced 'nan' in float array by 0.0, since at the end it is usually meaningful and convenient.

It comes with a slight performance improvement: https://github.com/thu-ml/tianshou/issues/96

output_V2

### CODE REVIEW

**Comment by duburcqa** [comment]
@Trinkle23897 Just for you to know, I'm an engineer and for the first time, and thanks to this PR (and all the previous ones) a real case application is finally running without raising an exception :laughing: Perhaps it is time to release 0.2.4 :stuck_out_tongue:

**Comment by Trinkle23897** [comment]
Sure and much appreciate~
How about the documentation update?

**Comment by duburcqa** [comment]
Yes I'll do it now. Could you please tell me what/where to update them and what you have in mind exactly ?

**Comment by Trinkle23897** [comment]
https://github.com/thu-ml/tianshou/blob/master/CONTRIBUTING.md

**Comment by duburcqa** [comment]
Yes I know this, but you mean only updating this page https://tianshou.readthedocs.io/en/latest/api/tianshou.data.html ?

**Comment by Trinkle23897** [comment]
Exactly. I mean that you can add the new feature in https://tianshou.readthedocs.io/en/latest/api/tianshou.data.html#tianshou.data.Batch
```
Batch has other methods, including __getitem__(), __len__(), append(), and split():
```

also with buffer.

**Comment by Trinkle23897** [comment]
You can make the documentation show `__xxx__` by modifying this line:
https://github.com/thu-ml/tianshou/blob/70aa7bf93ebe25d61e4566f3880b07a40eb702aa/docs/conf.py#L61
For example, add `__idiv__` into 'special-members'.
And the readthedocs page will be updated only if the master version has been updated. So it is better to preview the documentation in your own local computer.

**Comment by duburcqa** [comment]
Added missing documentation. It turns out that I didn't modified the behaviour of Buffer by any means, it was only internal modifications, so I have not updated it.

**Comment by Trinkle23897** [comment]
Could it support slicing method such as `batch[:, i]`?
Because I find that the advance RNN hidden state (#19) needs this and otherwise I should recursively get the value.

**Comment by duburcqa** [comment]
In practice it shouldn't be very hard, but conceptually Batch is missing the notion of multidimensional samples, for instance only `len` is available, and it corresponds to the dimension to the first axis of the stored data. The first step would be to add a `shape` method, similar to numpy. After that It shouldn't be much trouble. Maybe you could open an issue.

The issue is that `list` cannot be supported since computing the shape of a `list` is very expensive and grows exponentially with depth.

**Comment by duburcqa** [comment]
Adding this property to `Batch` class would do the trick (note that `isinstance` `if ... else` should be used to early return `[]` in case of native Python scalar) :

```
    @property
    def shape(self) -> List[int]:
        """Return self.size."""
        if len(self.__dict__.keys()) == 0:
            return []
        else:
            data_shape = []
            for v in self.__dict__.values():
                try:
                    data_shape.append(v.shape)
                except AttributeError:
                    raise TypeError("No support for 'shape' method with "\
                                    f"type {type(v)} in class Batch.")
            return min(*data_shape) if len(data_shape) > 1 else data_shape[0]
```

Then, `_valid_bounds(len(self), index)` must be replaced by `np.all(map(_valid_bounds, zip(self.shape, index))`, ONLY in the case where index is a actually a list of list of indices. No need to change anything else, since the rest of the method is already compatible with multidimensional indexing (expect for list, but list is not supported by `shape`, so an exception will be raised before reaching this point).

NB: You can ping me on the issue/PR if you want a review.

**Comment by duburcqa** [comment]
By the way, it may be good to overload direct assignment to make sure that every list are converted to numpy. Indeed, list are a recurrent issue for slicing, stacking and concatenation, and it introduces many `isinstance` `if ... else ...` for no real advantage.
